src/Orleans.Core/Messaging/ILegacyGatewayListProviderConfigurator.cs

```diff
@@ -0,0 +1,72 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Text;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Options;
+using Orleans.Configuration;
+using Orleans.Configuration.Options;
+using Orleans.Runtime.Configuration;
+using Orleans.Runtime;
+
+namespace Orleans.Messaging
+{
+    /// <summary>
+    /// LegacyGatewayProviderConfigurator configure GatewayListProvider in the legacy way, which is from ClientConfiguration
+    /// </summary>
+    public interface ILegacyGatewayListProviderConfigurator
```

Comment:
It need to be public so it can be accessible in the extension projects, where this interface is implemented. 
As for the doubts on making it public. I worry a little about it too. But since we will be announcing that we are supporting the legacy way to configure gateway and membership, so it make sense for users to understand the existence of this class if they discovered this class. And it is not very discoverable either.